### PR TITLE
fix: synchronize with typeorm_metadata table only if needed

### LINKED ISSUE #9173: relation "typeorm_metadata" does not exist while creating a view in migration
## Issue Description

This issue is probably related to [4923](https://github.com/typeorm/typeorm/issues/4923), pull request [4956](https://github.com/typeorm/typeorm/pull/4956) doesn't seem to work here.

### Expected Behavior

I try to implement a migration that creates views for postgres. Before that migration another migration installs all tables successfully (fresh installation). I'm running the migrations manually using the typeorm-cli.

I'm using the following code to create a view (just for testing before implementing the real query):

````typescript
 console.log(`Create view...`);
    await queryRunner.createView(new View({
      name: 'tool_view',
      materialized: false,
      expression: (connection: DataSource) => {
        return connection.createQueryBuilder(ToolEntity, 'tool').select('*');
      }
    }));
````

It should create a view that just represents the same table as "tool".

### Actual Behavior

I get the following error message:

````
QueryFailedError: relation "typeorm_metadata" does not exist
    at PostgresQueryRunner.query (/snapshot/project/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async PostgresQueryRunner.executeQueries (/snapshot/project/node_modules/typeorm/query-runner/BaseQueryRunner.js:421:13)
    at async PostgresQueryRunner.createView (/snapshot/project/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:475:9) {
  query: 'INSERT INTO "typeorm_metadata"("database", "schema", "table", "type", "name", "value") VALUES (DEFAULT, $1, DEFAULT, $2, $3, $4)',
  parameters: [
    'public',
    'VIEW',
    'test_view',
    'SELECT * FROM "tool" "tool"'
  ],
  driverError: error: relation "typeorm_metadata" does not exist
      at Parser.parseErrorMessage (/snapshot/project/node_modules/pg-protocol/dist/parser.js:287:98)
      at Parser.handlePacket (/snapshot/project/node_modules/pg-protocol/dist/parser.js:126:29)
      at Parser.parse (/snapshot/project/node_modules/pg-protocol/dist/parser.js:39:38)
      at Socket.<anonymous> (/snapshot/project/node_modules/pg-protocol/dist/index.js:11:42)
      at Socket.emit (node:events:527:28)
      at addChunk (node:internal/streams/readable:315:12)
      at readableAddChunk (node:internal/streams/readable:289:9)
      at Socket.Readable.push (node:internal/streams/readable:228:10)
      at TCP.onStreamRead (node:internal/stream_base_commons:190:23) {
    length: 115,
    severity: 'ERROR',
    code: '42P01',
    detail: undefined,
    hint: undefined,
    position: '13',
    internalPosition: undefined,
    internalQuery: undefined,
    where: undefined,
    schema: undefined,
    table: undefined,
    column: undefined,
    dataType: undefined,
    constraint: undefined,
    file: 'parse_relation.c',
    line: '1384',
    routine: 'parserOpenTable'
  },
  length: 115,
  severity: 'ERROR',
  code: '42P01',
  detail: undefined,
  hint: undefined,
  position: '13',
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'parse_relation.c',
  line: '1384',
  routine: 'parserOpenTable'
}
````

### Steps to Reproduce

1. Create a Datasource connection for PostgreSQL
2. Create a migration and implement code that creates any view.
3. Run the migration using the typeorm -cli.

### Possible solution

Create "typeorm_metadata" automatically next to "typeorm_migrations" table. That would ensure that the typeorm_metadata table exists.

### My Environment

| Dependency          | Version  |
| ---                 | ---      |
| Operating System           | MacOS  12.4  |
| Node.js version        |  v16.15.0 |
| Typescript version   |  v4.7.3 |
| TypeORM version    |  v.0.3.7 |
| Postgres   |  v14.2 |

### Relevant Database Driver(s)

| DB Type           | Reproducible  |
|-------------------| ---           |
| `aurora-mysql`    | no            |
| `aurora-postgres` | no            |
| `better-sqlite3`  | no            |
| `cockroachdb`     | no            |
| `cordova`         | no            |
| `expo`            | no            |
| `mongodb`         | no            |
| `mysql`           | no            |
| `nativescript`    | no            |
| `oracle`          | no            |
| `postgres`        | yes            |
| `react-native`    | no            |
| `sap`             | no            |
| `spanner`         | no            |
| `sqlite`          | no            |
| `sqlite-abstract` | no            |
| `sqljs`           | no            |
| `sqlserver`       | no            |

### Are you willing to resolve this issue by submitting a Pull Request?

 -  ✅ Yes, I have the time, and I know how to start.
 - ✖️ Yes, I have the time, but I don't know how to start. I would need guidance.
 - ✖️ No, I don’t have the time, but I can support (using donations) development.
 - ✖️ No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

### PR DESCRIPTION
**Pull request by julianpoemp**
### Description of change

Creating a view by calling `createQueryRunner().createView(...)` results in an error: `QueryFailedError: relation "typeorm_metadata" does not exist` (see #9173). This pull request fixes #9173 by creating the metadata table if it doesn't exist.

Fixes: #9173

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run format` to apply prettier formatting
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change N/A
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by nacholupotti** [comment]
HI, getting same error here. Will this PR be merged soon?

**Comment by julianpoemp** [comment]
@nacholupotti the PR is ready for merge, but it's waiting for a reviewer. @AlexMesser can you review it, please?

**Comment by christian-forgacs** [comment]
We've the same issue :(

**Comment by julianpoemp** [comment]
@christian-forgacs the fastest solution is to clone the branch `fix-missing-typeorm-metadata` branch from my fork and build the typeorm package from it. After that you get a .zip archive that you can copy to your project and install using npm. That's how I do it in my projects.

Unfortunately I don't know when and whether this PR is going to be merged or not.

File: src/migration/MigrationExecutor.ts [hunk 1]
```diff
@@ -172,14 +192,13 @@ export class MigrationExecutor {
...
+        // create migrations table if it's not created yet
         await this.createMigrationsTableIfNotExist(queryRunner)
 
-        // create the typeorm_metadata table if necessary
+        // create the typeorm_metadata table if it's not created yet
         const schemaBuilder = this.connection.driver.createSchemaBuilder()
-
         if (InstanceChecker.isRdbmsSchemaBuilder(schemaBuilder)) {
-            await schemaBuilder.createMetadataTableIfNecessary(queryRunner)
+            await schemaBuilder.createTypeormMetadataTable(queryRunner)
```
Comment (AlexMesser) [inline src/migration/MigrationExecutor.ts:201]: is there any issues with `createMetadataTableIfNecessary()` method? Metadata table required only for views and generated columns.

File: src/migration/MigrationExecutor.ts [hunk 2]
```diff
@@ -135,6 +148,13 @@ export class MigrationExecutor {
             this.queryRunner || this.connection.createQueryRunner()
         // create migrations table if its not created yet
         await this.createMigrationsTableIfNotExist(queryRunner)
+
+        // create typeorm_metadata table if it's not created yet
+        const schemaBuilder = this.connection.driver.createSchemaBuilder()
+        if (InstanceChecker.isRdbmsSchemaBuilder(schemaBuilder)) {
+            await schemaBuilder.createTypeormMetadataTable(queryRunner)
```
Comment (AlexMesser) [inline src/migration/MigrationExecutor.ts:155]: no need it here. This method just shows existing migrations.

File: src/migration/MigrationExecutor.ts [hunk 3]
```diff
@@ -87,6 +94,12 @@ export class MigrationExecutor {
         return this.withQueryRunner(async (queryRunner) => {
             await this.createMigrationsTableIfNotExist(queryRunner)
 
+            // create typeorm_metadata table if it's not created yet
+            const schemaBuilder = this.connection.driver.createSchemaBuilder()
+            if (InstanceChecker.isRdbmsSchemaBuilder(schemaBuilder)) {
+                await schemaBuilder.createTypeormMetadataTable(queryRunner)
```
Comment (AlexMesser) [inline src/migration/MigrationExecutor.ts:100]: no need it here. This method just shows executed migrations.

File: src/migration/MigrationExecutor.ts [hunk 4]
```diff
@@ -64,6 +64,13 @@ export class MigrationExecutor {
     public async executeMigration(migration: Migration): Promise<Migration> {
         return this.withQueryRunner(async (queryRunner) => {
             await this.createMigrationsTableIfNotExist(queryRunner)
+
+            // create typeorm_metadata table if it's not created yet
+            const schemaBuilder = this.connection.driver.createSchemaBuilder()
+            if (InstanceChecker.isRdbmsSchemaBuilder(schemaBuilder)) {
+                await schemaBuilder.createTypeormMetadataTable(queryRunner)
```
Comment (AlexMesser) [inline src/migration/MigrationExecutor.ts:71]: replace `createTypeormMetadataTable()` with `createMetadataTableIfNecessary()` and check if it resolves your issue.

File: test/github-issues/9173/issue-9173.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import "reflect-metadata"
+import {
+    closeTestingConnections,
+    createTestingConnections,
+    reloadTestingDatabases,
+} from "../../utils/test-utils"
+import { DataSource, Table } from "../../../src"
+import { View } from "../../../src/schema-builder/view/View"
+
+describe("github issues > #9173 missing typeorm_metadata", () => {
```
Comment (AlexMesser) [inline test/github-issues/9173/issue-9173.ts:10]: you can look at [this test](https://github.com/typeorm/typeorm/tree/master/test/github-issues/2693) to see how to test migrations

File: src/schema-builder/RdbmsSchemaBuilder.ts [hunk 6]
```diff
@@ -1182,7 +1182,7 @@ export class RdbmsSchemaBuilder implements SchemaBuilder {
     /**
      * Creates typeorm service table for storing user defined Views and generate columns.
      */
-    protected async createTypeormMetadataTable(queryRunner: QueryRunner) {
+    public async createTypeormMetadataTable(queryRunner: QueryRunner) {
```
Comment (AlexMesser) [inline src/schema-builder/RdbmsSchemaBuilder.ts:1185]: use `createMetadataTableIfNecessary()` instead and leave it `protected`

File: src/migration/MigrationExecutor.ts [hunk 7]
```diff
@@ -317,9 +336,15 @@ export class MigrationExecutor {
...
             this.queryRunner || this.connection.createQueryRunner()
 
-        // create migrations table if its not created yet
+        // create migrations table if it's not created yet
         await this.createMigrationsTableIfNotExist(queryRunner)
 
+        // create typeorm_metadata table if it's not created yet
+        const schemaBuilder = this.connection.driver.createSchemaBuilder()
+        if (InstanceChecker.isRdbmsSchemaBuilder(schemaBuilder)) {
+            await schemaBuilder.createTypeormMetadataTable(queryRunner)
```
Comment (AlexMesser) [inline src/migration/MigrationExecutor.ts:345]: replace `createTypeormMetadataTable()` with `createMetadataTableIfNecessary()`

**Comment by julianpoemp** [comment]
@AlexMesser I applied your suggestions from your review. Now my test doesn't succeeds any more, even if it just calls 

````typescript
await connection.createQueryRunner().createView(
                new View({
                    name: "test_view",
                    expression: "SELECT * FROM test_table",
                }),
            )
````

So this issue isn't one that just happens in migrations, but rather every time `createView()` is called. So I had a look at how this function works. I found out that the `insertTypeormMetadataSql` function  from `BaseQueryrunner.ts` doesn't checks if the meta data table already exists and fails.

So I think there are two options:

1. We check in `insertTypeormMetadataSql` if the meta data table exists. If not we create it.

2.  The `createMetadataTableIfNecessary()` function somehow recognizes that a metadata_table is needed in the future as soon as `insertTypeormMetadataSql` is called.

I tried to implement option 1, but because this function isn't async, I can't create the meta data table using the schema builder. I don't know how to implement option 2, too.

**Comment by pleerock** [comment]
The way `createMetadataTableIfNecessary` works - it checks if you have classes decorated by `@View` and if you do have - it knows it must create metadata table for you.

Looks like you don't have such classes (decorated by `@View`) and it means you don't generate migration based on changes in your code, but instead manually write them. Correct me if I'm wrong.

Option #1 doesn't work as you pointed.

Option #2 isn't implementable because there is no way to TypeORM to guess if you have views hard-coded.

Option #3 won't work as well because all these QueryRunner methods must be side-effect-less and they only generate SQL strings (by pushing into up/down queries), we do it because we generate migrations out of SQLs pushed into these arrays.

typeorm_metadata is needed only if you have classes decorated by `@View` / `@Entity`, because it uses the metadata to sync old view/entity class definition with new one. Since you don't have a class and instead creating view dynamically using static code in the migrations, sync won't ever work. So, we don't really need metadata in such case. 

I suggest to add additional boolean param for such cases:

```ts
async createView(view: View, syncWithMetadata: boolean = false): Promise<void> {
        const upQueries: Query[] = []
        const downQueries: Query[] = []
        upQueries.push(this.createViewSql(view))
        if (syncWithMetadata)
            upQueries.push(await this.insertViewDefinitionSql(view))
        downQueries.push(this.dropViewSql(view))
        if (syncWithMetadata)
            downQueries.push(await this.deleteViewDefinitionSql(view))
        await this.executeQueries(upQueries, downQueries)
    }
```

We can make it `false` by default and send `true` where we use it in TypeORM codebase to keep current behavior.

**Comment by julianpoemp** [comment]
@pleerock 

> Looks like you don't have such classes (decorated by @View) and it means you don't generate migration based on changes in your code, but instead manually write them. Correct me if I'm wrong.

Yes.

> typeorm_metadata is needed only if you have classes decorated by @View / @Entity, because it uses the metadata to sync old view/entity class definition with new one. Since you don't have a class and instead creating view dynamically using static code in the migrations, sync won't ever work. So, we don't really need metadata in such case.

I understand. So, a workaround would be to create a separate View class even if I don't use it.

> I suggest to add additional boolean param for such cases:

That is a really short and easy solution, very nice! But wouldn't we have to change all createView() functions for each QueryRunner and change all code occurrences that needs the old behaviour? If we set the default to true, the old behaviour would be the default (no code changes needed) and only in special cases like my one the param is set to false...

````typescript
await connection.createQueryRunner().createView(
                new View({
                    name: "test_view",
                    expression: "SELECT * FROM test_table",
                }),
                false
)
````
in order to prevent the synchronization with metadata_table.

If you think this is the best solution, I could implement and test it. Please let me know how you decide.

**Comment by pleerock** [comment]
> But wouldn't we have to change all createView() functions for each QueryRunner and change all code occurrences that needs the old behaviour? 

We only need to supply `true` in TypeORM codebase, where we actually need current behavior.

Cases where users were using code call of `createView` (and they had metadata table) were not valid - because they don't need metadata for code-created views.

**Comment by julianpoemp** [comment]
@pleerock I implemented it as you described. I ran the test and I tested it in my project, it works fine. Thank you! 😄

**Comment by pleerock** [comment]
tests seems to fail

**Comment by julianpoemp** [comment]
@pleerock most of the tests run successfully but not the cockroachdb one. I can't find the reason, the test fails with my build and with that one from the origin typeorm master repository. Every time it's another error, I don't know what's going on here.

````console
 1) "before each" hook for "should load data when additional condition used"


  582 passing (11m)
  14 pending
  1 failing

  1) query builder > joins
       "before each" hook for "should load data when additional condition used":
     Error: Timeout of 120000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/Users/ips/repos/typeorm-orig/typeorm/build/compiled/test/functional/query-builder/join/query-builder-joins.js)
      at listOnTimeout (node:internal/timers:559:17)
      at processTimers (node:internal/timers:502:7)
````

````console
 1) "before each" hook for "should not override already set properties"


  569 passing (16m)
  14 pending
  1 failing

  1) query builder > entity updation
       "before each" hook for "should not override already set properties":
     QueryFailedError: internal error: deadline below read timestamp is nonsensical; txn has would have no chance to commit. Deadline: 1663672502.269035525,0. Read timestamp: 1663672555.764900000,1 Previous Deadline: 1663672502.269035525,0.
      at CockroachQueryRunner.query (src/driver/cockroachdb/CockroachQueryRunner.ts:302:19)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at CockroachQueryRunner.commitTransaction (src/driver/cockroachdb/CockroachQueryRunner.ts:190:25)
      at CockroachQueryRunner.clearDatabase (src/driver/cockroachdb/CockroachQueryRunner.ts:2479:46)
      at DataSource.dropDatabase (src/data-source/DataSource.ts:356:17)
      at DataSource.synchronize (src/data-source/DataSource.ts:315:29)
      at async Promise.all (index 3)
````

**Comment by pleerock** [comment]
@julianpoemp nevermind cockroachdb is dumb and fails sometimes. This one is a good to merge. Thank you for contribution!

**Pull request merged**